servlet/src/main/java/com/linecorp/armeria/server/servlet/ServletHttpExchange.java

```diff
@@ -0,0 +1,160 @@
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.handler.codec.http.DefaultFullHttpRequest;
+import io.netty.handler.codec.http.HttpMethod;
+import io.netty.handler.codec.http.HttpVersion;
+import io.netty.util.AttributeKey;
+
+/**
+ * Servlet object (contains 3 big objects: request, response, TCP channel).
+ */
+public final class ServletHttpExchange {
+    private static final AttributeKey<ServletHttpExchange> CHANNEL_ATTR_KEY_EXCHANGE =
+            AttributeKey.valueOf(ServletHttpExchange.class + "#ServletHttpExchange");
+
+    private ServletHttpRequest request;
+    private ServletHttpResponse response;
+    private DefaultServiceRequestContext channelHandlerContext;
+    private ServletContextImpl servletContext;
+    private boolean isHttpKeepAlive;
+
+    private ServletHttpExchange() {
+    }
+
+    /**
+     * Get new instance.
+     */
+    public static ServletHttpExchange newInstance(ServletContextImpl servletContext,
+                                                  DefaultServiceRequestContext context,
+                                                  AggregatedHttpRequest request) {
+        final DefaultFullHttpRequest fullHttpRequest = parseHttpRequest(context, request);
+        final ServletHttpExchange instance = new ServletHttpExchange();
+        instance.servletContext = servletContext;
+        instance.channelHandlerContext = context;
+
+        //Create a new servlet request object
+        instance.request = ServletHttpRequest.newInstance(instance, fullHttpRequest);
+        //Create a new servlet response object
+        instance.response = ServletHttpResponse.newInstance(instance);
+        return instance;
+    }
+
+    /**
+     * Whether the pipe is active.
+     */
+    public static boolean isChannelActive(DefaultServiceRequestContext channelHandlerContext) {
+        return true;
+    }
+
+    /**
+     * Is http keep alive.
+     */
+    public boolean isHttpKeepAlive() {
+        return isHttpKeepAlive;
+    }
+
+    /**
+     * Get http servlet request.
+     */
+    public ServletHttpRequest getRequest() {
+        return request;
+    }
+
+    /**
+     * Get servlet context.
+     */
+    public ServletContextImpl getServletContext() {
+        return servletContext;
+    }
+
+    /**
+     * Get http servlet response.
+     */
+    public ServletHttpResponse getResponse() {
+        return response;
+    }
+
+    /**
+     * Get channel handler context.
+     */
+    public DefaultServiceRequestContext getChannelHandlerContext() {
+        return channelHandlerContext;
+    }
+
+    /**
+     * Get server address.
+     */
+    public InetSocketAddress getServerAddress() {
+        return servletContext.getServerAddress();
+    }
+
+    /**
+     * Get local address.
+     */
+    public InetSocketAddress getLocalAddress() {
+        final SocketAddress socketAddress = channelHandlerContext.localAddress();
+        if (socketAddress == null) {
```

Comment:
It's never `null`, so we don't need this check

Reply:
totally agree.